### PR TITLE
Feature/coh 61 question state date

### PR DESCRIPTION
**Pull request by Bishwin**

### CODE REVIEW

File: src/test/java/uk/gov/hmcts/reform/coh/controller/question/TestQuestionResponse.java [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+        ObjectMapper mapper = new ObjectMapper();
+        QuestionResponse request = mapper.readValue(json, QuestionResponse.class);
+
+        assertEquals("How do you do your shopping?", request.getQuestionHeaderText());
+        assertEquals("DRAFTED", request.getCurrentState().getName());
+        assertEquals("2018-07-06 11:44:42.526", request.getCurrentState().getDatetime());
+    }
+
+
+}
```
Comment (johnhosolirius) [inline src/test/java/uk/gov/hmcts/reform/coh/controller/question/TestQuestionResponse.java]: Needs a cucumber test to check that the current state has been returned from the endpoint

File: src/test/resources/json/question/standard_question_response.json [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+  "question_ordinal": "1",
+  "question_header_text": "How do you do your shopping?",
+  "question_body_text": "Provide a detailed description of the difficulty you face when doing your shopping",
+  "owner_reference": "owner_idam_reference_123",
+  "question_id": "b312022c-2f5b-4a43-9aa1-de45f075e415",
+  "current_question_state": {
+    "name": "DRAFTED",
+    "datetime": "2018-07-06 11:44:42.526"
+  }
+}
```
Comment (johnhosolirius) [inline src/test/resources/json/question/standard_question_response.json]: Doesn't match the API:
"current_question_state": {
        "state_name": "question_issued",
        "state_datetime": "5 February 2018 14:00 GMT"
      },

File: scripts/test/test-coverage.sh [hunk 3]
```diff
@@ -0,0 +1,7 @@
+#!/usr/bin/env bash
+
+./gradlew clean check jacocoTestReport --info
+
+xdg-open build/reports/jacoco/test/html/index.html
+open build/reports/jacoco/test/html/index.html
+start "" build/reports/jacoco/test/html/index.html
```
Comment (johnhosolirius) [inline scripts/test/test-coverage.sh]: No idea why Jake changed this

File: src/aat/java/uk/gov/hmcts/reform/coh/functional/bdd/steps/QuestionSteps.java [hunk 4]
```diff
@@ -133,6 +134,18 @@ public void get_all_questions_for_a_online_hearing() throws Throwable {
         }
     }
 
+    @And("^the get request is sent to retrieve the submitted question$")
```
Comment (johnhosolirius) [inline src/aat/java/uk/gov/hmcts/reform/coh/functional/bdd/steps/QuestionSteps.java]: This is a GET to retrieve a saved question

File: src/aat/java/uk/gov/hmcts/reform/coh/functional/bdd/steps/QuestionSteps.java [hunk 5]
```diff
@@ -261,4 +274,17 @@ public void the_response_contains_n_questions(int count) throws Throwable {
         AllQuestionsResponse questionResponses = mapper.readValue(rawJson, AllQuestionsResponse.class);
         assertEquals(count, questionResponses.getQuestions().size());
     }
+
+    @And("^the question id matches$")
```
Comment (johnhosolirius) [inline src/aat/java/uk/gov/hmcts/reform/coh/functional/bdd/steps/QuestionSteps.java]: This is to check that the returned Question ID is the same as the one requested in the GET request

File: src/main/java/uk/gov/hmcts/reform/coh/controller/question/QuestionResponseMapper.java [hunk 6]
```diff
@@ -9,11 +9,17 @@
...
     QUESTION_BODY_TEXT(Question::getQuestionText, QuestionRequest::setQuestionBodyText),
     OWNER_REFERENCE(Question::getOwnerReferenceId, QuestionRequest::setOwnerReference),
-    QUESTION_STATE(q -> {return q.getQuestionState().getState();}, (qr, s) -> { qr.getQuestionState().setState(s);});
+    QUESTION_STATE(q -> {return q.getQuestionState().getState();}, (qr, s) -> { qr.getCurrentState().setName(s);}),
+    STATE_TIME(q -> {
+        if (!q.getQuestionStateHistories().isEmpty()){
+            return q.getQuestionStateHistories().get(q.getQuestionStateHistories().size()-1).getDateOccurred().toString();
+        }
+        return null;
+        }, (qr, s) -> { qr.getCurrentState().setDatetime(s);});
```
Comment (johnhosolirius) [inline src/main/java/uk/gov/hmcts/reform/coh/controller/question/QuestionResponseMapper.java]: The state datetime is already sorted, so the last item in the list must be the latest

File: src/main/java/uk/gov/hmcts/reform/coh/service/QuestionRoundService.java [hunk 7]
```diff
@@ -145,7 +145,7 @@ public QuestionRound getQuestionRoundByRoundId(OnlineHearing onlineHearing, int
 
     public void issueQuestionRound(OnlineHearing onlineHearing, QuestionState questionState, int questionRoundNumber) {
         List<Question> questions = getQuestionsByQuestionRound(onlineHearing, questionRoundNumber);
-        questions.stream().forEach(q -> q.addState(questionState));
+        questions.stream().forEach(q -> q.setQuestionState(questionState));
         questions.stream().forEach(q -> questionRepository.save(q));
```
Comment (phillipDenness) [inline src/main/java/uk/gov/hmcts/reform/coh/service/QuestionRoundService.java]: Need to call updateQuestionStateHistory to set issue state in history aswell.

File: scripts/test/test-coverage.sh (hunk 3 above)
Comment (johnhosolirius) [reply scripts/test/test-coverage.sh]: Dunno mate. That xdg-open is uses file associated to open an application, so I'm guessing that Jake and maybe Rose will have something set up.

**Pull request merged**